vsintegration/src/FSharp.Editor/Diagnostics/UnusedDeclarationsAnalyzer.fs

```diff
@@ -29,80 +29,85 @@ type internal UnusedDeclarationsAnalyzer() =
             isEnabledByDefault = true,
             customTags = DiagnosticCustomTags.Unnecessary)
     
-    let symbolUseComparer =
-        { new IEqualityComparer<FSharpSymbolUse> with
-            member __.Equals (x, y) = 
-                if x.IsFromDefinition && y.IsFromDefinition then
-                    x.RangeAlternate = y.RangeAlternate
-                else
-                    x.Symbol.DeclarationLocation = y.Symbol.DeclarationLocation || x.Symbol.IsEffectivelySameAs y.Symbol
-            member __.GetHashCode x = 
-                x.Symbol.DeclarationLocation 
-                |> Option.orElseWith (fun _ -> x.Symbol.ImplementationLocation) 
-                |> hash }
+    let isPotentiallyUnusedDeclaration (symbol: FSharpSymbol) : bool =
+        match symbol with
+        // Determining that a record, DU or module is used anywhere requires inspecting all their enclosed entities (fields, cases and func / vals)
+        // for usages, which is too expensive to do. Hence we never gray them out.
+        | :? FSharpEntity as e when e.IsFSharpRecord || e.IsFSharpUnion || e.IsInterface || e.IsFSharpModule || e.IsClass -> false
+        // FCS returns inconsistent results for override members; we're skipping these symbols.
+        | :? FSharpMemberOrFunctionOrValue as f when 
+                f.IsOverrideOrExplicitInterfaceImplementation ||
+                f.IsBaseValue ||
+                f.IsConstructor -> false
+        // Usage of DU case parameters does not give any meaningful feedback; we never gray them out.
+        | :? FSharpParameter -> false
+        | _ -> true
 
-    let countSymbolsUses (symbolsUses: FSharpSymbolUse[]) =
-        let result = Dictionary<FSharpSymbolUse, int>(symbolUseComparer)
-    
-        for symbolUse in symbolsUses do
-            match result.TryGetValue symbolUse with
-            | true, count -> result.[symbolUse] <- count + 1
-            | _ -> result.[symbolUse] <- 1
-        result
+    let getUnusedDeclarationRanges (symbolsUses: FSharpSymbolUse[]) (isScript: bool) =
+        let definitions =
+            symbolsUses
+            |> Array.filter (fun su -> 
+                su.IsFromDefinition && 
```

Comment:
nitpick: this boolean logic might be best tucked away in a function `isValidForUnusedDeclaration` to ease readability